File: src/phone.ts
// UK-first phone validation for the booking form.
//
// A booking lives or dies on this number: the operator rings it to confirm the
// price and the driver rings it from the roadside. "At least seven digits" let
// through numbers nobody could ring back — a premium-rate 09, a mobile missing
// a digit, 1234567 — and the job arrived with no way to reach the customer.
// The backend applies the same rules (backend/app/phone.py), so a request that
// skips this form is held to them too.

export type PhoneCheck = { ok: true; formatted: string } | { ok: false; error: string };

/** Digits plus the punctuation people genuinely type into a phone number. */
const ALLOWED = /^[\d\s()+.-]+$/;

const EMERGENCY = new Set(['999', '112', '911']);

const fail = (reason: string): PhoneCheck => ({
  ok: false,
  error: `Please enter a valid phone number — ${reason}`,
});

/** 011x and 01x1 are the big-city codes (0113 Leeds, 0161 Manchester). */
const isCityCode = (d: string): boolean => d[2] === '1' || d[3] === '1';

/** "020 7946 0018", "0161 496 0000", "01234 567890": how each range is written. */
function formatLandline(d: string): string {
  if (d.startsWith('02')) return `${d.slice(0, 3)} ${d.slice(3, 7)} ${d.slice(7)}`;
  if (d.startsWith('03')) return `${d.slice(0, 4)} ${d.slice(4, 7)} ${d.slice(7)}`;
  if (d.length === 11 && isCityCode(d)) {
    return `${d.slice(0, 4)} ${d.slice(4, 7)} ${d.slice(7)}`;
  }
  return `${d.slice(0, 5)} ${d.slice(5)}`;
}

function lengthProblem(length: number): string {
  return length < 11
    ? 'it looks a digit or two short. UK numbers have 11 digits, like 07700 900123.'
    : 'it has too many digits. UK numbers have 11, like 07700 900123.';
}

/** A national-format UK number, leading 0 included. */
function checkUk(d: string): PhoneCheck {
  if (!d.startsWith('0')) return fail('UK numbers start with 0 or +44.');

  switch (d[1]) {
    case '7': {
      if (d.length !== 11) return fail(lengthProblem(d.length));
      // 070 is a "personal number" that forwards anywhere at a premium, and
      // 076 is pagers — except 07624, which is Isle of Man mobiles.
      if (d.startsWith('070') || (d.startsWith('076') && !d.startsWith('07624'))) {
        return fail("that isn't a mobile number we can ring.");
      }
      return { ok: true, formatted: `${d.slice(0, 5)} ${d.slice(5)}` };
    }
    case '1':
    case '2':
    case '3':
      // A handful of small 01 areas still have ten-digit numbers; the city
      // codes never do, so "0161 496 000" is a digit short, not a village.
      if (d.length === 11 || (d[1] === '1' && d.length === 10 && !isCityCode(d))) {
        return { ok: true, formatted: formatLandline(d) };
      }
      return fail(lengthProblem(d.length));
    case '8':
    case '9':
      return fail("we can't ring back an 08 or 09 number. A mobile or landline, please.");
    default:
      return fail('please use a UK mobile or landline.');
  }
}

export function checkPhone(input: string): PhoneCheck {
  const raw = input.trim();
  if (!raw) return fail('we need it so the driver can reach you.');
  if (!ALLOWED.test(raw)) return fail('use digits only.');
  if (raw.lastIndexOf('+') > 0) return fail('a + can only go at the start.');

  // "+44 (0)161 …" is how plenty of people write it; the (0) is never dialled.
  let digits = raw.replace(/\(0\)/g, '').replace(/\D/g, '');
  const international = raw.startsWith('+') || digits.startsWith('00');
  if (digits.startsWith('00')) digits = digits.slice(2);

  if (!international && EMERGENCY.has(digits)) {
    return {
      ok: false,
      error: 'If anyone is hurt or in danger, call 999 now. Otherwise enter your own number.',
    };
  }

  // UK in international form, with or without the +: back to national.
  if (digits.startsWith('44')) return checkUk(`0${digits.slice(2).replace(/^0/, '')}`);

  if (international) {
    // E.164 allows up to 15 digits; nothing real is shorter than 8.
    if (digits.length < 8 || digits.length > 15)
      return fail('that international number looks incomplete.');
    return { ok: true, formatted: `+${digits}` };
  }

  return checkUk(digits);
}

/** The tidy form of a valid number, or the input trimmed if it isn't one. */
export function formatPhone(input: string): string {
  const result = checkPhone(input);
  return result.ok ? result.formatted : input.trim();
}

File: backend/app/phone.py
"""UK-first phone validation for bookings. Mirrors src/phone.ts.

A booking is useless if nobody can ring the customer back, so the API holds
every submission to the same rules the form does — a request that skips the
form cannot store 1234567 or a premium-rate 09 number either. Valid numbers
come back in one tidy shape ("07700 900123", "0161 496 0000", "+33612345678")
so the operator reads them the same way every time.
"""

import re

_ALLOWED = re.compile(r"^[\d\s()+.\-]+$")
_EMERGENCY = {"999", "112", "911"}


def _is_city_code(d: str) -> bool:
    """011x and 01x1 are the big-city codes (0113 Leeds, 0161 Manchester)."""
    return d[2] == "1" or d[3] == "1"


def _format_landline(d: str) -> str:
    if d.startswith("02"):
        return f"{d[:3]} {d[3:7]} {d[7:]}"
    if d.startswith("03"):
        return f"{d[:4]} {d[4:7]} {d[7:]}"
    if len(d) == 11 and _is_city_code(d):
        return f"{d[:4]} {d[4:7]} {d[7:]}"
    return f"{d[:5]} {d[5:]}"


def _uk(d: str) -> str:
    if not d.startswith("0"):
        raise ValueError("UK phone numbers start with 0 or +44")
    kind = d[1] if len(d) > 1 else ""
    if kind == "7":
        if len(d) != 11:
            raise ValueError("UK mobile numbers have 11 digits")
        # 070 personal numbers and 076 pagers, except 07624 Isle of Man mobiles.
        if d.startswith("070") or (d.startswith("076") and not d.startswith("07624")):
            raise ValueError("not a mobile number that can be rung back")
        return f"{d[:5]} {d[5:]}"
    if kind in {"1", "2", "3"}:
        # A few small 01 areas still have ten-digit numbers; city codes never do.
        if len(d) == 11 or (kind == "1" and len(d) == 10 and not _is_city_code(d)):
            return _format_landline(d)
        raise ValueError("UK landline numbers have 11 digits")
    if kind in {"8", "9"}:
        raise ValueError("08 and 09 numbers cannot be rung back")
    raise ValueError("not a UK mobile or landline number")


def normalise_phone(value: str) -> str:
    """Return the tidy form of a callable number, or raise ValueError."""
    raw = value.strip()
    if not raw:
        raise ValueError("phone number is required")
    if not _ALLOWED.match(raw):
        raise ValueError("phone number may only contain digits, spaces and + ( ) - .")
    if raw.rfind("+") > 0:
        raise ValueError("+ can only appear at the start of a phone number")

    # "+44 (0)161 …": the (0) is never dialled.
    digits = re.sub(r"\D", "", raw.replace("(0)", ""))
    international = raw.startswith("+") or digits.startswith("00")
    if digits.startswith("00"):
        digits = digits[2:]

    if not international and digits in _EMERGENCY:
        raise ValueError("an emergency number is not a callback number")

    if digits.startswith("44"):
        return _uk("0" + re.sub(r"^0", "", digits[2:]))

    if international:
        if not 8 <= len(digits) <= 15:
            raise ValueError("international phone number looks incomplete")
        return f"+{digits}"

    return _uk(digits)
